jupyter_server/gateway/managers.py

```diff
@@ -376,6 +381,49 @@ class GatewayKernelManager(AsyncKernelManager):
     def _default_cache_ports(self):
         return False  # no need to cache ports here
 
+    # A list of pathlib objects, each pointing at an event
+    # schema to register with this kernel manager's eventlogger.
+    # This trait should not be overridden.
+
+    @property
+    def core_event_schema_paths(self) -> t.List[pathlib.Path]:
+        return [DEFAULT_EVENTS_SCHEMA_PATH / "kernel_actions" / "v1.yaml"]
+
+    # This trait is intended for subclasses to override and define
+    # custom event schemas.
+    extra_event_schema_paths = List(
+        default_value=[],
+        help="""
+        A list of pathlib.Path objects pointing at to register with
+        the kernel manager's eventlogger.
+        """,
+    ).tag(config=True)
+
+    event_logger = Instance(EventLogger)
+
+    @default("event_logger")
+    def _default_event_logger(self):
+        """Initialize the logger and ensure all required events are present."""
+        if self.parent is not None and hasattr(self.parent, "event_logger"):
+            logger = self.parent.event_logger
+        else:
+            # If parent does not have an event logger, create one.
+            logger = EventLogger()
+        # Ensure that all the expected schemas are registered. If not, register them.
+        schemas = self.core_event_schema_paths + self.extra_event_schema_paths
+        for schema_path in schemas:
+            # Try registering the event.
+            try:
+                logger.register_event_schema(schema_path)
+            # Pass if it already exists.
+            except SchemaRegistryException:
+                pass
+        return logger
+
+    def emit(self, schema_id, data):
+        """Emit an event from the kernel manager."""
+        self.event_logger.emit(schema_id=schema_id, data=data)
```

Comment:
I think we should derive `GatewayKernelManager` from `ServerKernelManager` (this should have been done when the latter got introduced a while back actually - my bad), then let inheritance bring in these attributes since these should be identical to those defined in `ServerKernelManager`.  
```suggestion
```

Reply:
Sounds good. I've updated it. 

I wasn't sure if we wanted to do this, since GatewayKernelManager inherited from `AsyncKernelManager` while ServerKernelManager inherits from `AsyncIOLoopKernelManager`. Will this be an issue?

Reply:
Thanks.  Yeah, I suspect `GatewayKernelManager` was incorrect all along and should have derived from `AsyncIOLoopKernelManager` from the get-go.  It really doesn't call `super()` anyway, essentially deriving method signatures (and attributes) and overriding those where necessary (e.g., `_default_cache_ports()`).

(Looking now, GKM does call `super().__init__()`, but AILKM does define that method, while AKM does, so we're not introducing any new code there.  We should be good.)